test/e2e/tests/eg_upgrade.go

```diff
@@ -0,0 +1,166 @@
+			routeNN := types.NamespacedName{Name: "http-backend-eg-upgrade", Namespace: ns}
+			gwNN := types.NamespacedName{Name: "same-namespace", Namespace: ns}
+			gwAddr := kubernetes.GatewayAndHTTPRoutesMustBeAccepted(t, suite.Client, suite.TimeoutConfig, suite.ControllerName, kubernetes.NewGatewayRef(gwNN), routeNN)
+			reqURL := url.URL{Scheme: "http", Host: http.CalculateHost(t, gwAddr, "http"), Path: "/backend-upgrade"}
+			t.Log("Attempting to upgrade the last version of eg deployment")
+			err = helmUpgradeChartFromPath(relName, depNS, "../../../charts/gateway-helm", t)
+			if err != nil {
+				t.Fatalf("Failed to upgrade the release: %s", err.Error())
+			}
+
+			// wait for everything to startup
+			kubernetes.NamespacesMustBeReady(t, suite.Client, suite.TimeoutConfig, namespaces)
+			loadSuccess := make(chan bool)
+			// can be used to abort the test after deployment restart is complete or failed
+			aborter := periodic.NewAborter()
+			t.Log("Starting load generation", "reqURL:", reqURL.String())
+
+			// Run load async and continue to restart deployment
+			go runLoadAndWait(t, suite.TimeoutConfig, loadSuccess, aborter, reqURL.String())
+			t.Log("Stopping load generation and collecting results")
+			aborter.Abort(false) // abort the load either way
+			// Wait for the goroutine to finish
+			result := <-loadSuccess
+			if !result {
+				t.Errorf("Load test failed")
+			}
+		})
+	},
+}
+
+func helmUpgradeChartFromPath(relName, relNamespace, chartPath string, t *testing.T) error {
+	actionConfig := new(action.Configuration)
+	if err := actionConfig.Init(cli.New().RESTClientGetter(), relNamespace, "secret", t.Logf); err != nil {
+		return err
+	}
+
+	// Set installation options.
+	upgrade := action.NewUpgrade(actionConfig)
+	upgrade.Namespace = relNamespace
+	upgrade.WaitForJobs = true
+
+	// Load the chart from a local directory.
+	chart, err := loader.Load(chartPath)
+	if err != nil {
+		return err
+	}
+
+	// Run the installation.
+	values := map[string]interface{}{
+		"deployment": map[string]interface{}{
+			"envoyGateway": map[string]interface{}{
+				"imagePullPolicy": "IfNotPresent",
+			},
+		},
+	}
+	_, err = upgrade.Run(relName, chart, values)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func helmInstall(relName, relNamespace string, tag string, t *testing.T) error {
+	actionConfig := new(action.Configuration)
+	if err := actionConfig.Init(cli.New().RESTClientGetter(), relNamespace, "secret", t.Logf); err != nil {
+		return err
+	}
+
+	// Set installation options.
+	install := action.NewInstall(actionConfig)
+	install.ReleaseName = relName
+	install.Namespace = relNamespace
+	install.CreateNamespace = true
+	install.Version = tag
+	install.WaitForJobs = true
+
+	registryClient, err := registry.NewClient()
+	if err != nil {
+		return err
+	}
+	install.SetRegistryClient(registryClient)
+	chartPath, err := install.LocateChart("oci://docker.io/envoyproxy/gateway-helm", cli.New())
+	if err != nil {
+		return err
+	}
```

Comment:
lets add a TODO comment in here saying we need to explicitly reinstall CRDs, and raise a separate GH issue to work on that as a follow up